### PR TITLE
SNS - Minor changes in GPIO implementation

### PR DESCRIPTION
**Pull request by licornes-fluos**
Change for GPIO protocol, matched IGPIO and HardwareGpio, and implements IGPIO in HardwareGpio.

### CODE REVIEW

File: lib/io/gpio.hpp [hunk 1]
```diff
@@ -34,8 +34,10 @@ class IGpioWriter {
  */
 class IGpio {
  public:
-  virtual std::optional<std::shared_ptr<IGpioReader>> getReader(const std::uint8_t pin) = 0;
-  virtual std::optional<std::shared_ptr<IGpioWriter>> getWriter(const std::uint8_t pin) = 0;
+  virtual std::optional<std::shared_ptr<IGpioReader>> getReader(const std::uint8_t pin,
+                                                                const Edge edge) = 0;
+  virtual std::optional<std::shared_ptr<IGpioWriter>> getWriter(const std::uint8_t pin,
+                                                                const Edge edge) = 0;
```
Comment (TomLonergan03) [inline lib/io/gpio.hpp:40]: this is the other thing, a bunch of stuff like `Edge` are defined in `hardware_gpio.hpp`, which i don't think should be there

File: lib/io/hardware_gpio.hpp [hunk 2]
```diff
@@ -51,7 +51,7 @@ class HardwareGpioWriter : public IGpioWriter {
  * be instantiated at the top level and then provided to users through the IGpio interface.
  * Ensure inputted pins are defined as pin = 32*X + Y (GPIOX_Y)
  */
-class HardwareGpio {
+class HardwareGpio : public IGpio{
```
Comment (TomLonergan03) [inline lib/io/hardware_gpio.hpp:54]:
```suggestion
class HardwareGpio : public IGpio {
```

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/io/gpio.hpp (hunk 1 above)
Comment (licornes-fluos) [reply lib/io/gpio.hpp:40]: Where would it be if not here though?

Comment (TomLonergan03) [reply lib/io/gpio.hpp:40]: yes, those should be defined here as this doesn't build

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]
`clang-format` + fixes in keyence then good to go

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]
`DummyGpio` will need updated to match the interface changes also

**Review by davidbeechey — APPROVED** [review]

**Review by TomLonergan03 — APPROVED** [review]

**Pull request merged**